pkg/deploy/controller/deployment/controller.go

```diff
@@ -56,9 +56,35 @@ func (c *DeploymentController) Handle(deployment *kapi.ReplicationController) er
 
 		deploymentPod, err := c.podClient.createPod(deployment.Namespace, podTemplate)
 		if err != nil {
-			// If the pod already exists, it's possible that a previous CreatePod succeeded but
-			// the deployment state update failed and now we're re-entering.
-			if !kerrors.IsAlreadyExists(err) {
+			if kerrors.IsAlreadyExists(err) {
+				// If the pod already exists, it's possible that a previous CreatePod succeeded but
+				// the deployment state update failed and now we're re-entering.
+				// Ensure that the pod is the one we created by verifying the annotation on it
+				existingPod, err := c.podClient.getPod(deployment.Namespace, deployutil.DeployerPodNameForDeployment(deployment))
+				if err != nil {
+					c.recorder.Eventf(deployment, "failedCreate", "Error getting existing deployer pod for %s: %v", deployutil.LabelForDeployment(deployment), err)
+					return fmt.Errorf("couldn't fetch existing deployer pod for %s: %v", deployutil.LabelForDeployment(deployment), err)
+				}
+				if deployutil.DeploymentNameFor(existingPod) == deployment.Name {
+					// we'll just set the deploymentPod so that pod name annotation
+					// can be set on the deployment below
+					deploymentPod = existingPod
```

Comment:
I agree that in principal the image should be checked and I have thought about it. However, currently, the image is really not the determining factor in tying a deployer pod to a deployment. For example, there are areas in the code that will update deployment status based on changes to the state/phase of the pod, regardless of the image that the pod is based on. 

As of now, I would highly recommend using annotations for the check.